src/registered.rs

```diff
@@ -0,0 +1,345 @@
+    /// value in [ZIP 32: Registered subtree root key generation][regroot].
+    ///
+    /// [regroot]: https://zips.z.cash/zip-0032#registered-subtree-root-key-generation
+    ///
+    /// # Panics
+    ///
+    /// Panics if:
+    /// - the context string is empty or longer than 252 bytes.
+    /// - the seed is shorter than 32 bytes or longer than 252 bytes.
+    fn master(context_string: &[u8], seed: &[u8]) -> Self {
+        with_ikm(context_string, seed, |ikm| Self {
+            inner: HardenedOnlyKey::master(ikm),
+        })
+    }
+
+    /// Derives a child key from a parent key at a given index and empty tag.
+    ///
+    /// This is a convenience function equivalent to
+    /// `self.derive_child_with_tag(index, &[])`.
+    pub fn derive_child(&self, index: ChildIndex) -> Self {
+        self.derive_child_with_tag(index, &[])
+    }
+
+    /// Derives a child key from a parent key at a given index and (possibly empty) tag.
+    ///
+    /// Defined in [ZIP 32: Registered child key derivation][regckd].
+    ///
+    /// [regckd]: https://zips.z.cash/zip-0032#registered-child-key-derivation
+    pub fn derive_child_with_tag(&self, index: ChildIndex, tag: &[u8]) -> Self {
+        Self {
+            inner: self.inner.derive_child_with_tag(index, tag),
+        }
+    }
+
+    /// Derives a 64-byte child cryptovalue from a parent key at a given index
+    /// and (possibly empty) tag.
+    ///
+    /// Defined in [ZIP 32: Full-width child cryptovalue derivation][fwccd].
+    ///
+    /// [fwccd]: https://zips.z.cash/zip-0032#full-width-child-cryptovalue-derivation
+    pub fn derive_child_cryptovalue(&self, index: ChildIndex, tag: &[u8]) -> [u8; 64] {
+        self.inner.ckdh_internal(index, 1, tag)
+    }
+
+    /// Returns the key material for this key.
+    pub fn data(&self) -> &[u8; 32] {
+        self.inner.parts().0
+    }
+
+    /// Returns the chain code for this key.
+    pub fn chain_code(&self) -> &ChainCode {
+        self.inner.parts().1
+    }
+}
+
+/// Derives a 64-byte cryptovalue (for use as key material for example), for a
+/// registered application protocol at the given non-empty subpath from the given
+/// seed. Each subpath element may consist of an index and (possibly empty) tag.
+///
+/// - `zip_number` is the number of the ZIP defining the application protocol.
+///   The corresponding hardened index (with empty tag) will be prepended to the
+///   `subpath` to obtain the ZIP 32 path.
+/// - `context_string` is an identifier for the context in which this key will be
+///   used. It must be globally unique.
+///
+/// # Panics
+///
+/// Panics if:
+/// - the context string is empty or longer than 252 bytes.
+/// - the seed is shorter than 32 bytes or longer than 252 bytes.
+/// - the subpath is empty.
+pub fn cryptovalue_from_subpath(
+    context_string: &[u8],
+    seed: &[u8],
+    zip_number: u16,
+    subpath: &[(ChildIndex, &[u8])],
```

Comment:
Same comment here as above: `subpath` should be documented and its elements should be values of a struct, not tuples.